pyccel/codegen/printing/ccode.py

```diff
@@ -762,7 +764,7 @@ def _print_ModuleHeader(self, expr):
                         class_scope.rename_function(func, f"{classDef.name}__{func.name.lstrip('__')}")
                         funcs += f"{self.function_signature(func)};\n"
             classes += "};\n"
-        funcs += '\n'.join(f"{self.function_signature(f)};" for f in expr.module.funcs)
+        funcs += '\n'.join(f"{self.function_signature(f)};" for f in expr.module.funcs if f.is_annotated and not f.is_inline)
```

Comment:
because we sometimes don't annotate the inlined function, but probably `is_inlined` is sufficient

Reply:
I agree, but why is this only a problem in the C printer?

Reply:
In C we call the method `function_signature` which will crash if the function is not annotated, but in Fortran, we go directly to `_print_FunctionDef` which returns an empty string

Reply:
Ok. So then this problem is common to all languages which print a header file. This means that it will also be needed when we print stub files for Python. Can the list be filtered in the semantic stage? It seems odd to me that we still have syntactic objects in the codegen stage

Reply:
In The semantics stage, it might be difficult, because we have the imports, so the function might be called in another file, so it will need to be annotated.
as for the Python case, it might be fine, because all the information needed for stubfiles is in the not annotated function. The issue with the Python Case is the printing of the not annotated node of the FunctionDef, there it raises an error, and I don't know how to fix it, or what should the behaviour be.

Reply:
You are right. The general version must be printed in the stub file and kept in the pickled file to ensure functions can import all versions.

> The issue with the Python Case is the printing of the not annotated node of the FunctionDef, it raises errors, and I don't know how to fix it, or what should the behavior be.

What errors are raised? Why is the `FunctionDef` not annotated?

Reply:
because the inlined function is not annotated if it is not used, so printing it will raise errors in the Python printer, because it expects the function to be annotated.

Reply:
So the main problem is that the Python code doesn't inline anything it just adds the decorator?

Reply:
What do we want stub files to contain for inline functions? I would expect their definition to be in the stub file. I believe the idea for stub files is that we can parse this file and have all information necessary to use the module. This means we avoid the parsing of the function def bodies. It is similar to a .h file in C. As we need to know about the body in order to inline the function I think the body of the inline function ought to be printed in the stub file. This is also similar to the handling of inline functions in C

Reply:
the issue with the Python printer is that we don't need the nodes to be annotated to print the function, but we in the printer require that otherwise it will crash in many places

Reply:
Ah that's true :thinking: that makes it trickier. I was hoping to use the Python printer to print these stub files.

I'm also not sure how useful the Python printer is. But it can be useful while onboarding new interns as it makes it easier to break up the PRs. It is mostly just used as a sanity check. Can we use unparse to handle this? We can get the original Python AST from `ast.basic.PyccelAstNode.python_ast`

Reply:
you want to use unparse to handle what exactly ? the stub files or use it instead of the printer?

Reply:
there is this library created by mypy https://mypy.readthedocs.io/en/stable/stubgen.html that generates stub files

Reply:
Something like:
```python
def _print_FunctionDef(self, expr):
    if expr.is_inline:
        return unparse(expr.python_ast)
        
    ....
```
That way this would work for the printer and in the stub files.

To be clear I am not sure that this is necessarily the best idea but I don't have an alternative idea currently

Reply:
We already add the annotations in the Python printer so I don't know if the stubgen would add much?

Reply:
> We already add the annotations in the Python printer so I don't know if the stubgen would add much?

yes not much, it just something that exists

Reply:
Yes. I don't think people should be inlining OpenMP loops though and in other cases it doesn't really matter if we lose the comments

Reply:
but I suspect unparse will raise an error if it encounters the `Comment` object

Reply:
Ah, I hadn't thought of this. Maybe we can brainstorm alternative solutions in the meeting

Reply:
I have tested it, it works without doing anything

Reply:
[311ea3f](https://github.com/pyccel/pyccel/pull/1728/commits/311ea3f86c9a98c1a91b5d36991341bfc15cbd52)